### ISSUE TITLE
 OpenShiftDeploymentVerifier build step fails with false positive.

### ISSUE THREAD

**Issue by arsogukpinar** [openshift/jenkins-plugin#11]
I have a build job with a OpenShiftDeploymentVerifier build step.  It is configured to verify if there is one pod running after build set. Even though I have one pod running this steps fails hence, my build job fails.

For this test following image has been used.
docker.io/openshift/jenkins-1-centos7 latest 212bac2225d0 6 days ago 655.4 MB
# Below you can find the relevant log extract.

OpenShiftDeploymentVerifier rc current count 1 rc desired count 1 step verification amount 1 current state Running
Auth - allowCertificate with incoming chain of len  2
Auth - allowCertificate returning true

OpenShiftDeploymentVerifier latest version:  19
Auth - allowCertificate with incoming chain of len  2
Auth - allowCertificate returning true

OpenShiftDeploymentVerifier current rc {
    "kind" : "ReplicationController",
    "apiVersion" : "v1",
    "metadata" : {
        "name" : "cakephp-example-19",
        "namespace" : "xyz",
        "selfLink" : "/api/v1/namespaces/xyz/replicationcontrollers/cakephp-example-19",
        "uid" : "49a7a46a-af9e-11e5-8b2c-fa163ef5b9d0",
        "resourceVersion" : "4043011",
        "generation" : 2,
        "creationTimestamp" : "2015-12-31T09:10:08Z",
        "labels" : {
            "openshift.io/deployment-config.name" : "cakephp-example",
            "template" : "cakephp-example"
        },
        "annotations" : {
            "kubectl.kubernetes.io/desired-replicas" : "0",
            "kubectl.kubernetes.io/update-source-id" : "cakephp-example-1:fb35e105-ae32-11e5-8b2c-fa163ef5b9d0",
            "openshift.io/deployer-pod.name" : "cakephp-example-19-deploy",
            "openshift.io/deployment-config.latest-version" : "19",
            "openshift.io/deployment-config.name" : "cakephp-example",
            "openshift.io/deployment.phase" : "Running",
            "openshift.io/encoded-deployment-config" : "{\"kind\":\"DeploymentConfig\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"cakephp-example\",\"namespace\":\"xyz\",\"selfLink\":\"/oapi/v1/namespaces/xyz/deploymentconfigs/cakephp-example\",\"uid\":\"3833352c-ae32-11e5-8b2c-fa163ef5b9d0\",\"resourceVersion\":\"4042975\",\"creationTimestamp\":\"2015-12-29T13:44:01Z\",\"labels\":{\"template\":\"cakephp-example\"},\"annotations\":{\"description\":\"Defines how to deploy the application server\"}},\"spec\":{\"strategy\":{\"type\":\"Rolling\",\"rollingParams\":{\"updatePeriodSeconds\":1,\"intervalSeconds\":1,\"timeoutSeconds\":600,\"maxUnavailable\":\"25%\",\"maxSurge\":\"25%\"},\"resources\":{}},\"triggers\":[{\"type\":\"ImageChange\",\"imageChangeParams\":{\"automatic\":true,\"containerNames\":[\"cakephp-example\"],\"from\":{\"kind\":\"ImageStreamTag\",\"name\":\"cakephp-example:latest\"},\"lastTriggeredImage\":\"172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac\"}},{\"type\":\"ConfigChange\"}],\"replicas\":1,\"selector\":{\"name\":\"cakephp-example\"},\"template\":{\"metadata\":{\"name\":\"cakephp-example\",\"creationTimestamp\":null,\"labels\":{\"name\":\"cakephp-example\"}},\"spec\":{\"containers\":[{\"name\":\"cakephp-example\",\"image\":\"172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac\",\"ports\":[{\"containerPort\":8080,\"protocol\":\"TCP\"}],\"env\":[{\"name\":\"DATABASE_SERVICE_NAME\"},{\"name\":\"DATABASE_ENGINE\"},{\"name\":\"DATABASE_NAME\"},{\"name\":\"DATABASE_USER\"},{\"name\":\"DATABASE_PASSWORD\"},{\"name\":\"CAKEPHP_SECRET_TOKEN\",\"value\":\"H3N7w8isnMAK6eNeTG8wSU7Fu3xoD6wbmWVEUZ68WkeVKNsrtU\"},{\"name\":\"CAKEPHP_SECURITY_SALT\",\"value\":\"seE36hj8kjfTtKmAeVs8qvtcv2BBQl2sAPmkDwwY\"},{\"name\":\"CAKEPHP_SECURITY_CIPHER_SEED\",\"value\":\"368014367837282015085010621150\"},{\"name\":\"OPCACHE_REVALIDATE_FREQ\",\"value\":\"2\"}],\"resources\":{},\"terminationMessagePath\":\"/dev/termination-log\",\"imagePullPolicy\":\"IfNotPresent\"}],\"restartPolicy\":\"Always\",\"terminationGracePeriodSeconds\":30,\"dnsPolicy\":\"ClusterFirst\",\"securityContext\":{}}}},\"status\":{\"latestVersion\":19,\"details\":{\"causes\":[{\"type\":\"ImageChange\",\"imageTrigger\":{\"from\":{\"kind\":\"DockerImage\",\"name\":\"172.30.209.107:5000/xyz/cakephp-example:latest\"}}}]}}}\n"
        }
    },
    "spec" : {
        "replicas" : 1,
        "selector" : {
            "deployment" : "cakephp-example-19",
            "deploymentconfig" : "cakephp-example",
            "name" : "cakephp-example"
        },
        "template" : {
            "metadata" : {
                "labels" : {
                    "deployment" : "cakephp-example-19",
                    "deploymentconfig" : "cakephp-example",
                    "name" : "cakephp-example"
                },
                "annotations" : {
                    "openshift.io/deployment-config.latest-version" : "19",
                    "openshift.io/deployment-config.name" : "cakephp-example",
                    "openshift.io/deployment.name" : "cakephp-example-19"
                }
            },
            "spec" : {
                "containers" : [{
                    "name" : "cakephp-example",
                    "image" : "172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac",
                    "ports" : [{
                        "containerPort" : 8080,
                        "protocol" : "TCP"
                    }],
                    "env" : [
                        {"name" : "DATABASE_SERVICE_NAME"},
                        {"name" : "DATABASE_ENGINE"},
                        {"name" : "DATABASE_NAME"},
                        {"name" : "DATABASE_USER"},
                        {"name" : "DATABASE_PASSWORD"},
                        {
                            "name" : "CAKEPHP_SECRET_TOKEN",
                            "value" : "H3N7w8isnMAK6eNeTG8wSU7Fu3xoD6wbmWVEUZ68WkeVKNsrtU"
                        },
                        {
                            "name" : "CAKEPHP_SECURITY_SALT",
                            "value" : "seE36hj8kjfTtKmAeVs8qvtcv2BBQl2sAPmkDwwY"
                        },
                        {
                            "name" : "CAKEPHP_SECURITY_CIPHER_SEED",
                            "value" : "368014367837282015085010621150"
                        },
                        {
                            "name" : "OPCACHE_REVALIDATE_FREQ",
                            "value" : "2"
                        }
                    ],
                    "terminationMessagePath" : "/dev/termination-log",
                    "imagePullPolicy" : "IfNotPresent"
                }],
                "restartPolicy" : "Always",
                "terminationGracePeriodSeconds" : 30,
                "dnsPolicy" : "ClusterFirst"
            }
        }
    },
    "status" : {
        "replicas" : 1,
        "observedGeneration" : 2
    }
}

OpenShiftDeploymentVerifier rc current count 1 rc desired count 1 step verification amount 1 current state Running
Auth - allowCertificate with incoming chain of len  2
Auth - allowCertificate returning true

OpenShiftDeploymentVerifier latest version:  19
Auth - allowCertificate with incoming chain of len  2
Auth - allowCertificate returning true

OpenShiftDeploymentVerifier current rc {
    "kind" : "ReplicationController",
    "apiVersion" : "v1",
    "metadata" : {
        "name" : "cakephp-example-19",
        "namespace" : "xyz",
        "selfLink" : "/api/v1/namespaces/xyz/replicationcontrollers/cakephp-example-19",
        "uid" : "49a7a46a-af9e-11e5-8b2c-fa163ef5b9d0",
        "resourceVersion" : "4043020",
        "generation" : 3,
        "creationTimestamp" : "2015-12-31T09:10:08Z",
        "labels" : {
            "openshift.io/deployment-config.name" : "cakephp-example",
            "template" : "cakephp-example"
        },
        "annotations" : {
            "kubectl.kubernetes.io/desired-replicas" : "0",
            "kubectl.kubernetes.io/update-source-id" : "cakephp-example-1:fb35e105-ae32-11e5-8b2c-fa163ef5b9d0",
            "openshift.io/deployer-pod.name" : "cakephp-example-19-deploy",
            "openshift.io/deployment-config.latest-version" : "19",
            "openshift.io/deployment-config.name" : "cakephp-example",
            "openshift.io/deployment.phase" : "Failed",
            "openshift.io/encoded-deployment-config" : "{\"kind\":\"DeploymentConfig\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"cakephp-example\",\"namespace\":\"xyz\",\"selfLink\":\"/oapi/v1/namespaces/xyz/deploymentconfigs/cakephp-example\",\"uid\":\"3833352c-ae32-11e5-8b2c-fa163ef5b9d0\",\"resourceVersion\":\"4042975\",\"creationTimestamp\":\"2015-12-29T13:44:01Z\",\"labels\":{\"template\":\"cakephp-example\"},\"annotations\":{\"description\":\"Defines how to deploy the application server\"}},\"spec\":{\"strategy\":{\"type\":\"Rolling\",\"rollingParams\":{\"updatePeriodSeconds\":1,\"intervalSeconds\":1,\"timeoutSeconds\":600,\"maxUnavailable\":\"25%\",\"maxSurge\":\"25%\"},\"resources\":{}},\"triggers\":[{\"type\":\"ImageChange\",\"imageChangeParams\":{\"automatic\":true,\"containerNames\":[\"cakephp-example\"],\"from\":{\"kind\":\"ImageStreamTag\",\"name\":\"cakephp-example:latest\"},\"lastTriggeredImage\":\"172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac\"}},{\"type\":\"ConfigChange\"}],\"replicas\":1,\"selector\":{\"name\":\"cakephp-example\"},\"template\":{\"metadata\":{\"name\":\"cakephp-example\",\"creationTimestamp\":null,\"labels\":{\"name\":\"cakephp-example\"}},\"spec\":{\"containers\":[{\"name\":\"cakephp-example\",\"image\":\"172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac\",\"ports\":[{\"containerPort\":8080,\"protocol\":\"TCP\"}],\"env\":[{\"name\":\"DATABASE_SERVICE_NAME\"},{\"name\":\"DATABASE_ENGINE\"},{\"name\":\"DATABASE_NAME\"},{\"name\":\"DATABASE_USER\"},{\"name\":\"DATABASE_PASSWORD\"},{\"name\":\"CAKEPHP_SECRET_TOKEN\",\"value\":\"H3N7w8isnMAK6eNeTG8wSU7Fu3xoD6wbmWVEUZ68WkeVKNsrtU\"},{\"name\":\"CAKEPHP_SECURITY_SALT\",\"value\":\"seE36hj8kjfTtKmAeVs8qvtcv2BBQl2sAPmkDwwY\"},{\"name\":\"CAKEPHP_SECURITY_CIPHER_SEED\",\"value\":\"368014367837282015085010621150\"},{\"name\":\"OPCACHE_REVALIDATE_FREQ\",\"value\":\"2\"}],\"resources\":{},\"terminationMessagePath\":\"/dev/termination-log\",\"imagePullPolicy\":\"IfNotPresent\"}],\"restartPolicy\":\"Always\",\"terminationGracePeriodSeconds\":30,\"dnsPolicy\":\"ClusterFirst\",\"securityContext\":{}}}},\"status\":{\"latestVersion\":19,\"details\":{\"causes\":[{\"type\":\"ImageChange\",\"imageTrigger\":{\"from\":{\"kind\":\"DockerImage\",\"name\":\"172.30.209.107:5000/xyz/cakephp-example:latest\"}}}]}}}\n"
        }
    },
    "spec" : {
        "replicas" : 0,
        "selector" : {
            "deployment" : "cakephp-example-19",
            "deploymentconfig" : "cakephp-example",
            "name" : "cakephp-example"
        },
        "template" : {
            "metadata" : {
                "labels" : {
                    "deployment" : "cakephp-example-19",
                    "deploymentconfig" : "cakephp-example",
                    "name" : "cakephp-example"
                },
                "annotations" : {
                    "openshift.io/deployment-config.latest-version" : "19",
                    "openshift.io/deployment-config.name" : "cakephp-example",
                    "openshift.io/deployment.name" : "cakephp-example-19"
                }
            },
            "spec" : {
                "containers" : [{
                    "name" : "cakephp-example",
                    "image" : "172.30.209.107:5000/xyz/cakephp-example@sha256:0a8132be3ff05f61d47bf5751b6dd1cb38054f7b208cfe6442821e3651f71cac",
                    "ports" : [{
                        "containerPort" : 8080,
                        "protocol" : "TCP"
                    }],
                    "env" : [
                        {"name" : "DATABASE_SERVICE_NAME"},
                        {"name" : "DATABASE_ENGINE"},
                        {"name" : "DATABASE_NAME"},
                        {"name" : "DATABASE_USER"},
                        {"name" : "DATABASE_PASSWORD"},
                        {
                            "name" : "CAKEPHP_SECRET_TOKEN",
                            "value" : "H3N7w8isnMAK6eNeTG8wSU7Fu3xoD6wbmWVEUZ68WkeVKNsrtU"
                        },
                        {
                            "name" : "CAKEPHP_SECURITY_SALT",
                            "value" : "seE36hj8kjfTtKmAeVs8qvtcv2BBQl2sAPmkDwwY"
                        },
                        {
                            "name" : "CAKEPHP_SECURITY_CIPHER_SEED",
                            "value" : "368014367837282015085010621150"
                        },
                        {
                            "name" : "OPCACHE_REVALIDATE_FREQ",
                            "value" : "2"
                        }
                    ],
                    "terminationMessagePath" : "/dev/termination-log",
                    "imagePullPolicy" : "IfNotPresent"
                }],
                "restartPolicy" : "Always",
                "terminationGracePeriodSeconds" : 30,
                "dnsPolicy" : "ClusterFirst"
            }
        }
    },
    "status" : {
        "replicas" : 0,
        "observedGeneration" : 3
    }
}

BUILD STEP EXIT: OpenShiftDeploymentVerifier deployment cakephp-example-19 failed
Build step 'Check Deployment Success in OpenShift' marked build as failure
# Finished: FAILURE

**Comment by gabemontero** [comment]
First - thanks for the interest in the plug-in.  And for providing the debug data for this issue.

Looking at the data, I cannot agree with your assessment, though if the intent of the deployment verifier was not clear to you, we can work to improve the documentation.

In particular, Generation 2 of the replication controller has the "openshift.io/deployment.phase" annotation set to "Running", but the latest Generation, Generation 3, has the "openshift.io/deployment.phase" set to "Failed".  The deployment verifier considers only the latest/current deployment.

I'll wait for your response to my findings before dispensing with this issue.

@bparees @mfojtik - FYI.

**Comment by bparees** [comment]
it does look like the doc leaves out mentioning that it checks that the latest deployment succeeded, so it'd be good to add that to the list of things that are checked.

```
"Check deployment success in OpenShift": determines whether the expected set of DeploymentConfig's, ReplicationController's, and active replicas are present based on prior use of either the "Scale deployments in OpenShift" (2) or "Trigger a deployment in OpenShift" (3) steps; its activities specifically include:

it first confirms the specified deployment config exists.
it then gets the list of all replication controllers for that DC, and finds the latest incarnation.
and then sees if within a reasonable time the current replica count is at least equal to the desired replica count.
```

**Comment by gabemontero** [comment]
I've updated the README with some further clarification and pushed to the repo.

Going ahead and closing this one out.

**Comment by arsogukpinar** [comment]
Hi Gabe,
Sorry for late response. Last couple of days I have been traveling. After your first comment, I did further testing. It took me sometime  before I can figure out the root cause of the issue.  It turns out that issue I was having was related to https://github.com/openshift/origin/issues/5787. 

My build job had "Scale deployments in Openshift" step. Before performing a build I was scaling down my app to 0 PODs and due to the issue I have pointed above rolling updates and validation step were failing.

It wasn't easy to spot but  while my replication controller with highest version was in "Failed" status and had 1 running POD, rolling updates were using the last replication controller with "Complete" status, which had 0 PODs active. This was triggering the issue above.  After I manually started an additional  POD with latest RC in "Complete" status and removed my scale down 0  step from my job, validation step was working as expected.
